lib/resty/openidc.lua

```diff
@@ -959,20 +973,88 @@ end
 local function openidc_load_jwt_and_verify_crypto(opts, jwt_string, asymmetric_secret,
 symmetric_secret, expected_algs, ...)
   local r_jwt = require("resty.jwt")
-  local enc_hdr, enc_payload, enc_sign = string.match(jwt_string, '^(.+)%.(.+)%.(.*)$')
-  if enc_payload and (not enc_sign or enc_sign == "") then
-    local jwt = openidc_load_jwt_none_alg(enc_hdr, enc_payload)
+  local jwt_obj
+
+  -- Expect a JWT encoded as a JWE, extracting parts
+  local part1, part2, part3, part4, part5 = string.match(jwt_string, '^(.+)%.(.+)%.(.+)%.(.+)%.(.*)$')
+  
+  -- No parts extracted, try to extract parts of a JWS encoded JWT
+  if not part1 and not part2 then
+    part1, part2, part3= string.match(jwt_string, '^(.+)%.(.+)%.(.*)$')
+    part4, part5 = nil
+  end
+
+  log(DEBUG, "part 1 : ",part1, ", part 2 : ",part2, ", part 3 : ",part3, ", part 4 : ",part4, ", part 5 : ",part5)
+  -- Determine type of JWT (simple JWT, JWS, JWE) :
+
+  -- Case : is a simple JWT
+  if part1 and not part2 and not part3  and not part4 and not part5 then
+    return nil, "token is not secured, it's a simple JWT."
+
+  -- Case : is an unsigned JWS
+  elseif part1 and part2 and (not part3 or part3 == "") and not part4 and not part5 then
+    -- part1 = JOSE Header, part2 = Payload, part3 = Signature, others are unused
+    local jwt = openidc_load_jwt_none_alg(part1, part2)
     if jwt then
       if opts.accept_none_alg then
         log(DEBUG, "accept JWT with alg \"none\" and no signature")
         return jwt
       else
         return jwt, "token uses \"none\" alg but accept_none_alg is not enabled"
       end
-    end -- otherwise the JWT is invalid and load_jwt produces an error
+    else
+      -- Return error when token look like a JWT or the token is unsigned but shouldn't (alg other than \"none\"")
+      return nil, "invalid unsigned jwt"
+    end 
+
+  -- Case : is a signed JWS
+  elseif part1 and part2 and part3 and not part4 and not part5 then
+    -- part1 = JOSE Header, part2 = Payload, part3 = Signature, others are unused
+    jwt_obj = r_jwt:load_jwt(jwt_string, nil)
+
+  -- Case : is a JWE, without or with preshared key
+  elseif (part1 and part2 and part3 and  part4 and not part5) or 
+          (part1 and part2 and part3 and  part4 and part5) then
+    -- part1 = JOSE Header, part2 = Initialization Vector, part3 = Cyphertext, part4 = Authentication Tag , others are unused
+    -- or 
+    -- part1 = JOSE Header, part2 = Pre-shared key, part3 = Initialization Vector, part4 = Cyphertext, part5 = Authentication Tag 
+    local jwe_header = cjson.decode(unb64(part1))
```

Comment:
shouldn't that be urlsafe base64 decoding, i.e. `openidc_base64_url_decode`? Also it may be good to use cjson_s and check for an error.